python/cuda_parallel/cuda/parallel/experimental/_bindings.pyx

```diff
@@ -0,0 +1,1664 @@
+            )
+        self.type_info.size = size
+        self.type_info.alignment = alignment
+        self.type_info.type = <cy_cccl_type_enum> type_enum.value
+
+    @property
+    def size(self):
+        return self.type_info.size
+
+    @property
+    def alignment(self):
+        return self.type_info.alignment
+
+    @property
+    def typenum(self):
+        return self.type_info.type
+
+    def as_bytes(self):
+        cdef uint8_t[:] mem_view = bytearray(sizeof(self.type_info))
+        memcpy(&mem_view[0], &self.type_info, sizeof(self.type_info))
+        return bytes(mem_view)
+
+
+cdef class Value:
+    cdef uint8_t[::1] state_obj
+    cdef TypeInfo value_type
+    cdef cy_cccl_value_t value_data;
+
+    def __cinit__(self, TypeInfo value_type, uint8_t[::1] state):
+        self.state_obj = state
+        self.value_type = value_type
+        self.value_data.type = value_type.type_info
+        self.value_data.state = <void *>&state[0]
+
+    @property
+    def type(self):
+        return self.value_type
+
+    property state:
+        def __get__(self):
+            return self.state_obj
+
+        def __set__(self, uint8_t[::1] new_value):
+            if (len(self.state_obj) == len(new_value)):
+                self.state_obj = new_value
+                self.value_data.state = <void *>&self.state_obj[0]
+            else:
+                raise ValueError("Size mismatch")
+
+    def as_bytes(self):
+        cdef uint8_t[:] mem_view = bytearray(sizeof(self.value_data))
+        memcpy(&mem_view[0], &self.value_data, sizeof(self.value_data))
+        return bytes(mem_view)
+
+
+cdef void ensure_buffer(object o) except *:
+    if not PyObject_CheckBuffer(o):
+        raise TypeError(
+            "Object with buffer protocol expected, "
+            f"got {type(o)}"
+        )
+
+
+cdef void * get_buffer_pointer(object o, size_t *size):
+    cdef int status = 0
+    cdef void *ptr = NULL
+    cdef Py_buffer view
+
+    status = PyObject_GetBuffer(o, &view, PyBUF_SIMPLE | PyBUF_ANY_CONTIGUOUS)
+    if status != 0:  # pragma: no cover
+        size[0] = 0
+        raise RuntimeError(
+            "Can not access simple contiguous buffer"
+        )
+
+    ptr = view.buf
+    size[0] = <size_t>view.len
+    PyBuffer_Release(&view)
+
+    return ptr
+
+
+cdef void * ctypes_typed_pointer_payload_ptr(object ctypes_typed_ptr):
+    "Get pointer to the value buffer represented by ctypes.pointer(ctypes_val)"
+    cdef size_t size = 0
+    cdef size_t *ptr_ref = NULL
+    ensure_buffer(ctypes_typed_ptr)
+    ptr_ref = <size_t *>get_buffer_pointer(ctypes_typed_ptr, &size)
+    return <void *>(ptr_ref[0])
+
+
+cdef void * ctypes_value_ptr(object ctypes_cdata):
+    "Get pointer to the value buffer behind ctypes_val"
+    cdef size_t size = 0
+    ensure_buffer(ctypes_cdata)
+    return get_buffer_pointer(ctypes_cdata, &size)
+
+
+cdef inline void * int_as_ptr(size_t ptr_val):
+    return <void *>(ptr_val)
+
+
+cdef class IteratorStateView:
```

Comment:
I don't think I fully understand the need for the following:

- `IteratorStateView`
- `IteratorState`,
- `Pointer`

I'm trying to understand: can we simplify by assuming that an iterator state is some Python object that is coercible either to a `memoryview` or [typedmemoryview](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#python-buffer-support)?